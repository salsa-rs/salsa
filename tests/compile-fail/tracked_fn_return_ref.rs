use salsa::Database as Db;

#[salsa::input]
struct MyInput {
    #[returns(ref)]
    text: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
struct ContainsRef<'db> {
    text: &'db str,
}

#[salsa::tracked]
fn tracked_fn_return_ref<'db>(db: &'db dyn Db, input: MyInput) -> &'db str {
    input.text(db)
}

#[salsa::tracked]
fn tracked_fn_return_struct_containing_ref<'db>(
    db: &'db dyn Db,
    input: MyInput,
) -> ContainsRef<'db> {
    ContainsRef {
        text: input.text(db),
    }
}

#[salsa::tracked]
fn tracked_fn_return_struct_containing_ref_elided_implicit<'db>(
    db: &'db dyn Db,
    input: MyInput,
) -> ContainsRef {
    ContainsRef {
        text: input.text(db),
    }
}

#[salsa::tracked]
fn tracked_fn_return_struct_containing_ref_elided_explicit<'db>(
    db: &'db dyn Db,
    input: MyInput,
) -> ContainsRef<'_> {
    ContainsRef {
        text: input.text(db),
    }
}

fn main() {}
