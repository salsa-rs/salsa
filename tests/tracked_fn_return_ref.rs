use salsa::Database;

#[salsa::input]
struct Input {
    number: usize,
}

#[salsa::tracked(returns(ref))]
fn test(db: &dyn salsa::Database, input: Input) -> Vec<String> {
    (0..input.number(db)).map(|i| format!("test {i}")).collect()
}

#[test]
fn invoke() {
    salsa::DatabaseImpl::new().attach(|db| {
        let input = Input::new(db, 3);
        let x: &Vec<String> = test(db, input);
        expect_test::expect![[r#"
            [
                "test 0",
                "test 1",
                "test 2",
            ]
        "#]]
        .assert_debug_eq(x);
    })
}
