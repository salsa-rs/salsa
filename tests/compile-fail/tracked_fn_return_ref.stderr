error[E0106]: missing lifetime specifier
  --> tests/compile-fail/tracked_fn_return_ref.rs:33:6
   |
33 | ) -> ContainsRef {
   |      ^^^^^^^^^^^ expected named lifetime parameter
   |
help: consider using the `'db` lifetime
   |
33 | ) -> ContainsRef<'db> {
   |                 +++++

warning: elided lifetime has a name
  --> tests/compile-fail/tracked_fn_return_ref.rs:33:6
   |
30 | fn tracked_fn_return_struct_containing_ref_elided_implicit<'db>(
   |                                                            --- lifetime `'db` declared here
...
33 | ) -> ContainsRef {
   |      ^^^^^^^^^^^ this elided lifetime gets resolved as `'db`
   |
   = note: `#[warn(elided_named_lifetimes)]` on by default

warning: elided lifetime has a name
  --> tests/compile-fail/tracked_fn_return_ref.rs:43:18
   |
40 | fn tracked_fn_return_struct_containing_ref_elided_explicit<'db>(
   |                                                            --- lifetime `'db` declared here
...
43 | ) -> ContainsRef<'_> {
   |                  ^^ this elided lifetime gets resolved as `'db`

error: lifetime may not live long enough
  --> tests/compile-fail/tracked_fn_return_ref.rs:15:67
   |
15 | fn tracked_fn_return_ref<'db>(db: &'db dyn Db, input: MyInput) -> &'db str {
   |                          --- lifetime `'db` defined here          ^ requires that `'db` must outlive `'static`

error: lifetime may not live long enough
  --> tests/compile-fail/tracked_fn_return_ref.rs:23:6
   |
20 | fn tracked_fn_return_struct_containing_ref<'db>(
   |                                            --- lifetime `'db` defined here
...
23 | ) -> ContainsRef<'db> {
   |      ^^^^^^^^^^^ requires that `'db` must outlive `'static`

error: lifetime may not live long enough
  --> tests/compile-fail/tracked_fn_return_ref.rs:43:6
   |
40 | fn tracked_fn_return_struct_containing_ref_elided_explicit<'db>(
   |                                                            --- lifetime `'db` defined here
...
43 | ) -> ContainsRef<'_> {
   |      ^^^^^^^^^^^ requires that `'db` must outlive `'static`
